sdk/core/core-rest-pipeline/src/policies/bearerTokenAuthenticationPolicy.ts

```diff
@@ -188,26 +216,96 @@ export function bearerTokenAuthenticationPolicy(
       try {
         response = await next(request);
       } catch (err: any) {
-        error = err;
-        response = err.response;
+        if (isRestError(error)) {
+          response = err.response;
+          error = err;
+        } else {
+          throw err;
+        }
       }
 
-      if (
-        callbacks.authorizeRequestOnChallenge &&
-        response?.status === 401 &&
-        getChallenge(response)
-      ) {
-        // processes challenge
-        const shouldSendRequest = await callbacks.authorizeRequestOnChallenge({
-          scopes: Array.isArray(scopes) ? scopes : [scopes],
-          request,
-          response,
-          getAccessToken,
-          logger,
-        });
-
-        if (shouldSendRequest) {
-          return next(request);
+      async function handleCaeChallenge(claims: string): Promise<PipelineResponse> {
+        const shouldSendRequest = await authorizeRequestOnCaeChallenge(
+          {
+            scopes: Array.isArray(scopes) ? scopes : [scopes],
+            response,
+            request,
+            getAccessToken,
+            logger,
+          },
+          claims,
+        );
+        // Send updated request and handle response for RestError
+        try {
+          if (shouldSendRequest) {
+            error = undefined;
+            response = await next(request);
+          }
+        } catch (err: any) {
```

Comment:
Do we have any tests that verify that we handle `next` throwing an error (a rejection)? i.e. coverage for these cases:
- `next` throws a `RestError` that contains a 401 challenge response that the policy can handle, in which case we should handle it. There might be multiple flavors of that (CAE challenge, some other kind of challenge)
- `next` throws a `RestError` that contains a 401 response this policy _cannot_ handle, in which case we re-throw the error
- `next` throws something that isn't a `RestError`, in which case we also re-throw that error

Reply:
I think the first one is covered in the test scenarios

Reply:
Good catch, I haven't added test for RestError and regular error. Will do!